libs/community/langchain_community/document_loaders/sqlalchemy.py

```diff
@@ -0,0 +1,144 @@
+import functools
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import sqlalchemy as sa
+
+from langchain_community.docstore.document import Document
+from langchain_community.document_loaders.base import BaseLoader
+
+
+class SQLAlchemyLoader(BaseLoader):
+    """
+    Load documents by querying database tables supported by SQLAlchemy.
+    Each document represents one row of the result.
+    """
+
+    def __init__(
+        self,
+        query: Union[str, sa.Select],
+        url: str,
+        page_content: Optional[Callable] = None,
+        metadata: Optional[Callable] = None,
+        source_columns: Optional[List[str]] = None,
+        include_rownum_into_metadata: bool = False,
+        include_query_into_metadata: bool = False,
+        sqlalchemy_kwargs: Optional[Dict] = None,
+    ):
+        """
+
+        Args:
+            query: The query to execute.
+            url: The SQLAlchemy connection string of the database to connect to.
+            page_content: Function to convert a row into a string to use as the
+              `page_content` of the document.
+              It is an optional keyword argument. By default, it serializes the
+              whole row into a string, including all columns.
+            metadata: Function to convert a row into a dictionary to use as the
+              `metadata` of the document.
+              It is an optional keyword argument. By default, no columns are
+              selected into the metadata dictionary.
+            source_columns: The names of the columns to use as the `source` within the
+              metadata dictionary. Optional.
+            include_rownum_into_metadata: Whether to include the row number into the
+              metadata dictionary. Optional. Default: False.
+            include_query_into_metadata: Whether to include the query expression into
+              the metadata dictionary. Optional. Default: False.
+            sqlalchemy_kwargs: More keyword arguments for SQLAlchemy's `create_engine`.
+        """
+        self.query = query
+        self.url = url
+        self.page_content = page_content
+        self.metadata = metadata
+        self.source_columns = source_columns
+        self.include_rownum_into_metadata = include_rownum_into_metadata
+        self.include_query_into_metadata = include_query_into_metadata
+        self.sqlalchemy_kwargs = sqlalchemy_kwargs or {}
+
+    def load(self) -> List[Document]:
```

Comment:
I think the sqlalchemy resultset is an iterator. So it would be more powerful to implement `lazy_load`

Reply:
Thank you for educating me about `lazy_load`. Implemented per 8920e2f80.